### DISCUSSION TITLE
How do SVG fonts make it into the final PDF?

### DISCUSSION THREAD

**Discussion by markkimsal** [dompdf/dompdf#2768] [Q&A]
I had an issue where my production server was using a bold font in an embedded SVG and a normal weight font on my local system.

After lots of digging, I did chmod a+w on the whole vendor dir then the server started using the same not-bold font inside the embedded SVG.

Now i'm kicking myself because I'm not sure if it was the SVG library that needed to write some font information or dompdf.  The lib/fonts/ folder looks no different on the server compared to my local dir, the cache-dist file is 4013 bytes.

It's very interesting to me why the embedded SVG used the last referenced font when some cache folder was not writable.  When looking at the 2 PDFs in a desktop PDF program, the server one (the one that used the bold font) had an extra Times New Roman font listed in the properties.

My question is: How are fonts referenced between the two drawing systems?  If I want to use Calibri or Georgia in a portion of an SVG, does the CPDF surface need to swap the font context in PDF?  Does it need to reference (or create a reference) to a new PDF font in the PDF document?

To control the font, I can embed one in dompdf, then have an empty span with an nbsp and a css class that uses the desired font right before the <img src="data:svg/base64,....".  But that's only for the whole SVG.

![image](https://user-images.githubusercontent.com/54099/152798675-419e3504-442b-48b7-946d-bee152e5f226.png)

**Comment by bsweeney** [comment]
php-svg-lib utilizes the Cpdf backend also utilized by Dompdf, meaning the state of that backend can impact how the SVG is rendered. When using the Cpdf backend, if a font is not found the last active font will be used. Which sounds like what you're seeing, generally, with the font being loaded.

php-svg-lib doesn't have a font management system like Dompdf, so you have to make sure the fonts are already available when it's processing a SVG document. You can bootstrap font loading mechanism using Dompdf (e.g. via the `@font-face` declaration). You didn't provide a sample of your HTML+CSS or SVG, but the write access issues might be due to using `@font-face` in your HTML document to reference fonts. There are two things to consider related to processing a new font...

First, font installation. When a font is processed by Dompdf it is parsed for font metrics and the font and the metrics file are written to the font directory. By default this is located in dompdf/lib/fonts. You can change where installed fonts are stored by modifying the fontDir Dompdf option:

```
$dompdf = new Dompdf([ "fontDir" => "/path/to/writeable/directory" ])
```

Second, font caching. With Cpdf, part of the process of selecting a font is generating a cache of the font metrics. This cache file is written to the directory specified by the fontCache Dompdf option, which is the same as the fontDir option when Dompdf is instantiated. You can change this value same as you would for the fontDir option.

---

The real issue is in how php-svg-lib handles fonts generally. Whereas Dompdf has a nicely integrated font management system, php-font-lib does not. This means that Dompdf will map the font name to the font file, but php-svg-lib expects you to specify the font file directly.

As an example let's say we're using the DejaVu Sans font bundled with Dompdf. In Dompdf you would reference it using `font-family: DejaVu Sans`. But in php-svg-lib you would reference the path to the actual font file, e.g. `font-family: /path/to/dompdf/lib/fonts/DejaVuSans`. If you're referencing a font dynamically loaded from a remote systems things get a little more complicated since a hash is appended to the filename to help prevent naming collisions.

As a final note, php-svg-lib doesn't support font fallback so you can only specify a single font name in your styling.
